std/random.d

```diff
@@ -1934,6 +1934,177 @@ A singleton instance of the default random number generator
+                    wenforce (CryptAcquireContextA(&_hProvider, null, null, PROV_RSA_FULL,
+                    CRYPT_NEWKEYSET | CRYPT_SILENT), "Could not acquire crypt context");
+                }
+                else
+                {
+                    wenforce (false, "Could not acquire crypt context");
+                }
+            }
+
+            popFront();
+        }
+
+    public:
+        enum bool isUniformRandom = true;
+        enum empty = false;
+
+        ulong front;
+
+        void popFront() @trusted nothrow @nogc
+        {
+            if (!CryptGenRandom(_hProvider, front.sizeof, cast(PBYTE) &front))
+                assert(0, "CryptGenRandom failed");
+        }
+
+        // TLS data is not destructed anyway right now...
+        version (none)
+        {
+            ~this()
+            {
+                CryptReleaseContext(_hProvider, 0);
+            }
+        }
+    }
+}
+
+version (SecureARC4Random)
+{
+    struct ARC4Random
+    {
+    private:
+        void initialize() @nogc @trusted nothrow
+        {
+            popFront();
+        }
+
+    public:
+        enum bool isUniformRandom = true;
+        enum empty = false;
+
+        ulong front;
+
+        void popFront() @safe nothrow @nogc
+        {
+            arc4random_buf(&front, front.sizeof);
+        }
+    }
+}
+
+version (UseFileRandom)
+{
+    import core.stdc.stdio : fclose, feof, ferror, fopen, fread;
+    alias IOType = typeof(fopen("a", "b"));
+
+    struct FileRandom
+    {
+    private:
+        IOType _file;
+
+        void initialize() @trusted
+        {
+            _file = fopen("/dev/urandom", "r");
+            if (!_file)
+                throw new Exception("Could not open /dev/urandom");
+            popFront();
+        }
+
+    public:
+        enum bool isUniformRandom = true;
+        enum empty = false;
+
+        ulong front;
+
+        void popFront() @trusted nothrow
+        {
+            auto res = fread(&front, 1, front.sizeof, _file);
+            if (res != front.sizeof)
+                assert(0, "Could not read from /dev/urandom");
+        }
+
+        // TLS data is not destructed anyway right now...
+        version (none)
+        {
+            ~this()
+            {
+                fclose(_file);
+            }
+        }
+    }
+}
+
+version (Windows)
+    alias OSRandom = WindowsRandom;
+else version (SecureARC4Random)
+    alias OSRandom = ARC4Random;
+else version (UseFileRandom)
+    alias OSRandom = FileRandom;
+else
+    static assert(false, "Platform not supported");
+
+/+
+Interface to the OS cryptographically secure random number generator.
+It is allocated per-thread on first use, to avoid synchronization overhead.
+
+Note:
+This RNG may block initially, if the OS did not yet collect enough entropy
+to return secure random numbers.
+
+Returns:
+A singleton instance of the operating system secure random number generator
++/
+@property ref OSRandom osRNDGen() @safe
```

Comment:
What about defining a buffer size?

```suggestion
@property ref OSRandom osRNDGen(size_t bufferSize = 256) @safe
```

Reply:
Ah, good idea. Though this lazy initialization is weird then, as the parameter will be evaluated only on the first call. There's also the problem that the initialization can actually cause more / different errors than fetching the data. So splitting these calls would make sense, but having initialization functions is not really phobos style...

Alternatively, we could not provide a default instance at all, but I think for this use case, we really want a thread-wide default.